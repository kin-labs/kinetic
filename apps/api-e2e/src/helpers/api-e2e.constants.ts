import { getAuthUsers } from '@kin-kinetic/api/config/data-access'
import { UserRole } from '@prisma/client'

const users = getAuthUsers(process.env.AUTH_USERS)
const { username, password } = users.find((user) => user.role === UserRole.Admin)

export const ADMIN_USERNAME = username
export const ADMIN_PASSWORD = password
