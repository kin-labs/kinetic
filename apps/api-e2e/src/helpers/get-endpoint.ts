import { INestApplication } from '@nestjs/common'
import * as request from 'supertest'

export function getEndpoint(app: INestApplication, endpoint: string) {
  return request(app.getHttpServer()).get(endpoint)
}
