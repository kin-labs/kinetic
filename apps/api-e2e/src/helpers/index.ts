export * from './api-e2e.constants'
export * from './get-endpoint'
export * from './initialize-e2e-app'
export * from './run-query'
export * from './random-int'
