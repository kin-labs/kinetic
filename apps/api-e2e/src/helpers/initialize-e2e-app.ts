import { INestApplication, ValidationPipe } from '@nestjs/common'
import { Test, TestingModule } from '@nestjs/testing'
import { AppModule } from '@kin-kinetic/api-app-module'

export async function initializeE2eApp(): Promise<INestApplication> {
  const moduleFixture: TestingModule = await Test.createTestingModule({
    imports: [AppModule],
  }).compile()

  const app = moduleFixture.createNestApplication()
  app.useGlobalPipes(new ValidationPipe({ transform: true }))
  app.setGlobalPrefix('api')
  await app.init()
  return app
}
