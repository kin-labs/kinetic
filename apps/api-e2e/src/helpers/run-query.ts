import { INestApplication } from '@nestjs/common'
import { UserRole } from '@prisma/client'
import { ASTNode } from 'graphql'
import { print } from 'graphql/language/printer'
import * as request from 'supertest'
import { Login, LoginInput } from '../generated/api-sdk'
import { ADMIN_PASSWORD } from './api-e2e.constants'

export function runGraphQLQuery(
  app: INestApplication,
  query: ASTNode,
  variables?: Record<string, unknown>,
  { token }: { token?: string } = {},
) {
  return request(app.getHttpServer())
    .post('/graphql')
    .set('Authorization', token ? `Bearer ${token}` : null)
    .send({ query: print(query), variables: variables })
}

export function runGraphQLQueryAdmin(
  app: INestApplication,
  token: string,
  query: ASTNode,
  variables?: Record<string, unknown>,
) {
  return runGraphQLQuery(app, query, variables, { token })
}

export function runLoginQuery(app: INestApplication, email: string, role: UserRole) {
  const input: LoginInput = { password: ADMIN_PASSWORD, email }
  return runGraphQLQuery(app, Login, { input }).expect(200)
}
