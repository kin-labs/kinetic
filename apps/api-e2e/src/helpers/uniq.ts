import { getRandomInt } from './random-int'

export function uniq(prefix: string) {
  return `${prefix}${uniqInt()}`
}
export function uniqInt(): number {
  return Math.floor(Math.random() * 10000000)
}
export function randomAppIndex(): number {
  return getRandomInt(1_000, 1_000_000)
}
