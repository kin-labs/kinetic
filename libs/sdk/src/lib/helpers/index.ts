export * from './get-solana-rpc-endpoint'
export * from './kin-to-quarks'
export * from './parse-kinetic-sdk-config'
export * from './parse-kinetic-sdk-endpoint'
