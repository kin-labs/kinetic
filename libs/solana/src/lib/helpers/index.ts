export * from './get-public-key'
export * from './parse-endpoint'
export * from './deserialize-transaction'
