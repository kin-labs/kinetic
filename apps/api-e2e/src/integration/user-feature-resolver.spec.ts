import { INestApplication } from '@nestjs/common'
import { Response } from 'supertest'
import {
  CreateUser,
  DeleteUser,
  UpdateUser,
  User,
  UserCreateInput,
  UserRole,
  Users,
  UserUpdateInput,
} from '../generated/api-sdk'
import { ADMIN_EMAIL, initializeE2eApp, runGraphQLQuery, runGraphQLQueryAdmin, runLoginQuery } from '../helpers'
import { uniq } from '../helpers/uniq'

function expectUnauthorized(res: Response) {
  expect(res).toHaveProperty('text')
  const errors = JSON.parse(res.text).errors
  expect(errors[0].message).toEqual('Unauthorized')
}

function randomUsername(): string {
  return uniq('user-')
}

describe('User (e2e)', () => {
  let app: INestApplication
  let userId: string | undefined
  let username: string | undefined
  let token: string | undefined

  beforeAll(async () => {
    app = await initializeE2eApp()
    const res = await runLoginQuery(app, ADMIN_EMAIL, UserRole.Admin)
    token = res.body.data.login.token
    username = randomUsername()
  })

  afterAll(async () => {
    userId = undefined
    username = undefined
    return app.close()
  })

  describe('Expected usage', () => {
    describe('CRUD', () => {
      it('should create a user', async () => {
        const name = uniq('user')
        const input: UserCreateInput = {
          email: `${name}@example.com`,
          password: 'password',
          username,
          name,
        }

        return runGraphQLQueryAdmin(app, token, CreateUser, { input })
          .expect(200)
          .expect((res) => {
            expect(res).toHaveProperty('body.data')
            const data = res.body.data?.created
            userId = data.id
            expect(data.username).toEqual(username)
            expect(data.email).toEqual(input.email)
            expect(data.name).toEqual(input.name)
            expect(data.role).toEqual(UserRole.User)
          })
      })

      it('should update a user', async () => {
        const input: UserUpdateInput = {
          name: `User ${username} edited`,
          role: UserRole.Admin,
          avatarUrl: 'test-avatar',
        }

        return runGraphQLQueryAdmin(app, token, UpdateUser, { userId, input })
          .expect(200)
          .expect((res) => {
            expect(res).toHaveProperty('body.data')
            const data = res.body.data?.updated

            expect(data.avatarUrl).toEqual('test-avatar')
            expect(data.username).toEqual(username)
            expect(data.name).toEqual(input.name)
            expect(data.role).toEqual(UserRole.Admin)
          })
      })

      it('should find a user', async () => {
        return runGraphQLQueryAdmin(app, token, User, { userId })
          .expect(200)
          .expect((res) => {
            expect(res).toHaveProperty('body.data')
            const data = res.body.data?.item

            expect(data.id).toEqual(userId)
            expect(data.username).toEqual(username)
          })
      })

      it('should find a list of users', async () => {
        return runGraphQLQueryAdmin(app, token, Users)
          .expect(200)
          .expect((res) => {
            expect(res).toHaveProperty('body.data')
            const data = res.body.data?.items

            expect(data.length).toBeGreaterThan(1)
            expect(data.find((user) => user.username === username)).toBeDefined()
          })
      })

      it('should delete a user', async () => {
        return runGraphQLQueryAdmin(app, token, DeleteUser, { userId })
          .expect(200)
          .expect((res) => {
            expect(res).toHaveProperty('body.data')
            const data = res.body.data?.deleted

            expect(data.id).toEqual(userId)
            expect(data.username).toEqual(username)
          })
      })
    })
  })

  describe('Unexpected usage', () => {
    describe('CRUD Constraints', () => {
      it('should not create a user with existing email', async () => {
        const username = randomUsername()
        const input: UserCreateInput = {
          username,
          password: 'password',
          email: uniq('email-'),
          name: `User ${username}`,
        }

        // Run once
        await runGraphQLQueryAdmin(app, token, CreateUser, { input })
        // Try to add again
        return runGraphQLQueryAdmin(app, token, CreateUser, { input })
          .expect(200)
          .expect((res) => {
            expect(res).toHaveProperty('error')
            const errors = JSON.parse(res.text).errors
            expect(errors[0].message).toEqual(`User with email ${input.email} already exists`)
          })
      })

      it('should not create a user with existing username', async () => {
        const username = randomUsername()
        const input: UserCreateInput = {
          username,
          password: 'password',
          email: uniq('email-'),
          name: `User ${username}`,
        }

        // Run once
        await runGraphQLQueryAdmin(app, token, CreateUser, { input })
        // Try to add again
        return runGraphQLQueryAdmin(app, token, CreateUser, { input: { ...input, email: uniq('email-') } })
          .expect(200)
          .expect((res) => {
            expect(res).toHaveProperty('error')
            const errors = JSON.parse(res.text).errors
            expect(errors[0].message).toEqual(`User with username ${input.username} already exists`)
          })
      })

      it('should not update a user that does not exist', async () => {
        const userId = uniq('user-')
        const input: UserUpdateInput = { name: `User ${userId}` }

        return runGraphQLQueryAdmin(app, token, UpdateUser, { userId, input })
          .expect(200)
          .expect((res) => {
            expect(res).toHaveProperty('error')
            const errors = JSON.parse(res.text).errors
            expect(errors[0].message).toEqual(`User with id ${userId} does not exist.`)
          })
      })

      it('should not find a user that does not exist', async () => {
        const userId = uniq('user-')
        return runGraphQLQueryAdmin(app, token, User, { userId })
          .expect(200)
          .expect((res) => {
            expect(res).toHaveProperty('error')
            const errors = JSON.parse(res.text).errors
            expect(errors[0].message).toEqual(`User with id ${userId} does not exist.`)
          })
      })

      it('should not delete a user that does not exist', async () => {
        const userId = uniq('user-')
        return runGraphQLQueryAdmin(app, token, DeleteUser, { userId })
          .expect(200)
          .expect((res) => {
            expect(res).toHaveProperty('error')
            const errors = JSON.parse(res.text).errors
            expect(errors[0].message).toEqual(`User with id ${userId} does not exist.`)
          })
      })
    })

    describe('Mall-formed Input', () => {
      it('should not create a user', async () => {
        const input: UserCreateInput = { username: undefined, email: undefined, password: undefined }

        return runGraphQLQuery(app, CreateUser, { input })
          .expect(400)
          .expect((res) => {
            expect(res).toHaveProperty('error')
            const errors = JSON.parse(res.text).errors
            expect(errors).toMatchSnapshot()
          })
      })

      it('should not update a user', async () => {
        const input: UserUpdateInput = { name: undefined }

        return runGraphQLQuery(app, UpdateUser, { userId: undefined, input })
          .expect(400)
          .expect((res) => {
            expect(res).toHaveProperty('error')
            const errors = JSON.parse(res.text).errors
            expect(errors).toMatchSnapshot()
          })
      })

      it('should not find a user', async () => {
        return runGraphQLQuery(app, User, { userId: undefined })
          .expect(400)
          .expect((res) => {
            expect(res).toHaveProperty('error')
            const errors = JSON.parse(res.text).errors
            expect(errors).toMatchSnapshot()
          })
      })

      it('should not delete a user', async () => {
        return runGraphQLQuery(app, DeleteUser, { userId: undefined })
          .expect(400)
          .expect((res) => {
            expect(res).toHaveProperty('error')
            const errors = JSON.parse(res.text).errors
            expect(errors).toMatchSnapshot()
          })
      })
    })

    describe('Unauthenticated Access', () => {
      it('should not create a user', async () => {
        const input: UserCreateInput = { username: username, password: 'password', email: uniq('email-') }

        return runGraphQLQuery(app, CreateUser, { input })
          .expect(200)
          .expect((res) => {
            expect(res).toHaveProperty('text')
            const errors = JSON.parse(res.text).errors
            expect(errors[0].message).toEqual('Unauthorized')
          })
      })

      it('should not update a user', async () => {
        const input: UserUpdateInput = { name: `User ${username} edited` }

        return runGraphQLQuery(app, UpdateUser, { userId, input })
          .expect(200)
          .expect((res) => expectUnauthorized(res))
      })

      it('should not find a user', async () => {
        return runGraphQLQuery(app, User, { userId })
          .expect(200)
          .expect((res) => expectUnauthorized(res))
      })

      it('should not find a list of users', async () => {
        return runGraphQLQuery(app, Users)
          .expect(200)
          .expect((res) => expectUnauthorized(res))
      })

      it('should not delete a user', async () => {
        return runGraphQLQuery(app, DeleteUser, { userId })
          .expect(200)
          .expect((res) => expectUnauthorized(res))
      })
    })
  })
})
