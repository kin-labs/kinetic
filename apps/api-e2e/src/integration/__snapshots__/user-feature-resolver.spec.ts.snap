// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User (e2e) Unexpected usage Mall-formed Input should not create a user 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "BAD_USER_INPUT",
    },
    "locations": Array [
      Object {
        "column": 21,
        "line": 1,
      },
    ],
    "message": "Variable \\"$input\\" got invalid value {}; Field \\"email\\" of required type \\"String!\\" was not provided.",
  },
  Object {
    "extensions": Object {
      "code": "BAD_USER_INPUT",
    },
    "locations": Array [
      Object {
        "column": 21,
        "line": 1,
      },
    ],
    "message": "Variable \\"$input\\" got invalid value {}; Field \\"password\\" of required type \\"String!\\" was not provided.",
  },
]
`;

exports[`User (e2e) Unexpected usage Mall-formed Input should not delete a user 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "BAD_USER_INPUT",
    },
    "locations": Array [
      Object {
        "column": 21,
        "line": 1,
      },
    ],
    "message": "Variable \\"$userId\\" of required type \\"String!\\" was not provided.",
  },
]
`;

exports[`User (e2e) Unexpected usage Mall-formed Input should not find a user 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "BAD_USER_INPUT",
    },
    "locations": Array [
      Object {
        "column": 12,
        "line": 1,
      },
    ],
    "message": "Variable \\"$userId\\" of required type \\"String!\\" was not provided.",
  },
]
`;

exports[`User (e2e) Unexpected usage Mall-formed Input should not update a user 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "BAD_USER_INPUT",
    },
    "locations": Array [
      Object {
        "column": 21,
        "line": 1,
      },
    ],
    "message": "Variable \\"$userId\\" of required type \\"String!\\" was not provided.",
  },
]
`;
